### PR TITLE
Flexible tlogs

### PR DESCRIPTION
**Pull request by toolslive**

### CODE REVIEW

File: src/client/remote_nodestream.ml [hunk 1]
```diff
@@ -83,12 +83,19 @@ class remote_nodestream ((ic,oc) as conn) =
...
+      let _should_reinit = ref false in
       let rec loop_parts () =
         Llio.input_int ic >>= function
         | (-2) -> Logger.info_f_ "loop_parts done"
         | 1 ->
           begin
             Logger.debug_f_ "loop_entries" >>= fun () ->
+            begin
+              if !_should_reinit
+              then tlog_coll # reinit()
```
Comment (domsj) [inline src/client/remote_nodestream.ml]: Why is this needed? Looks like it should be a concern of the tlogcollection itself?

File: src/client/remote_nodestream.ml [hunk 2]
```diff
@@ -101,17 +108,18 @@ class remote_nodestream ((ic,oc) as conn) =
...
-            tlog_coll # save_tlog_file name length ic >>= fun () ->
-            loop_parts ()
-          end
+           begin
+             let () = _should_reinit := true in
+             Logger.debug_f_ "save_file" >>= fun () ->
+             Llio.input_string ic >>= fun name ->
+             Llio.input_int64 ic >>= fun length ->
+             Logger.info_f_ "got %s (%Li bytes)" name length >>= fun () ->
+             tlog_coll # save_tlog_file name length ic >>= fun () ->
```
Comment (domsj) [inline src/client/remote_nodestream.ml]: the name used to save this tlog file comes from the other side ... which may differ from the name that's needed here
there's no guarantee the rollover points (and thus tlog names for a certain i) will be perfectly in sync
in theory it should even be possible to run the different nodes with different policies for when the rollover should occur?
(that would actually be a good test imo)

File: src/client/remote_nodestream.ml (hunk 1 above)
Comment (toolslive) [inline src/client/remote_nodestream.ml]: the `new_entry` function keeps track of when a roll over is needed, but it doesn't know that whole files were injected meanwhile. It's indeed possible to push this down and make it part of the state management as a side effect of `save_tlog_file`. I'll do that.

File: src/client/remote_nodestream.ml (hunk 2 above)
Comment (toolslive) [inline src/client/remote_nodestream.ml]: Yes. We should be more careful here. I'll adapt.

File: src/node/catchup_test.ml [hunk 3]
```diff
@@ -137,9 +137,11 @@ let teardown () =
   Logger.debug_ "end of teardown"
 
 let _tic filler_function n name verify_store =
-  Tlogcommon.tlogEntriesPerFile := 101;
+  let open Tlog_map in
+  let node_id = "node_name" in
+  TlogMap.make ~tlog_max_entries:101 _dir_name _tlx_dir node_id ~check_marker:true >>= fun tlog_map ->
```
Comment (domsj) [inline src/node/catchup_test.ml]: Did you add the construction of the tlogmap here so you could base some asserts on it?

File: src/node/collapser.ml [hunk 4]
```diff
@@ -120,13 +128,19 @@ let collapse_until (type s) (tlog_coll:Tlogcollection.tlog_collection)
...
-
-                          cb (Sn.of_int (first_tlog + !processed)) >>= fun () ->
-                          Lwt.return( processed := !processed + 1 )
+                          begin
+                            Logger.info_f_ "rollover_point pi:%Li next_rollover_point =%s"
+                                          pi (Sn.string_of !next_rollover_point)
+                            >>= fun () ->
+                            cb (first_tlog + !processed) >>= fun () ->
+                            let () = incr processed in
+                            let () = next_rollover_point := get_next_point (Sn.succ pi) in
```
Comment (domsj) [inline src/node/collapser.ml]: `let () = next_rollover_point := get_next_point pi in`
should work too? and stays correct even when you have a tlog with only 1 entry?

Also this got me thinking: are entries with the same i supposed to always be in the same tlog? If not what's the implication for the tlogmap and usages of # next_rollover ?

Comment (toolslive) [inline src/node/collapser.ml]: if you have: I as last entry of tlog_{n}, and I also as first entry in tlog_{n+1}, then the tlogmap will start folding at tlog_{n + 1} when starting at I, which is ok.

**Comment by domsj** [comment]
Catchup from an old arakoon to a newer one using flexible rollover will give issues ...
So it should either be blocked or fixed (by only streaming entries to older nodes).
This means you can put a new last_entries variant next to it and use that opportunity to also send along information regarding start/end i of a tlog?

File: src/main/replay_main.ml [hunk 5]
```diff
@@ -37,28 +37,35 @@ let replay_tlogs tlog_dir tlf_dir db_name end_i =
...
-                     | None -> Sn.start
-                     | Some e -> Sn.succ (Tlogcommon.Entry.i_of e)
-                 in
-                 Lwt.return i
-               end
-             | Some i -> Lwt.return i
-         end
+        TlogMap.make tlog_dir tlx_dir "node_name???" ~check_marker:false
+        >>= fun (tlog_map,_,_) ->
+        S.make_store ~lcnum:1024 ~ncnum:512 db_name >>= fun store ->
```
Comment (domsj) [inline src/main/replay_main.ml:45]: Hmm, this shouldn't use hardcoded values for lcnum/ncnum, it should come from the config.
This behaviour hasn't changed though, so just noticing this is something we may want to pick up sometime in the future

File: src/client/client_protocol.ml [hunk 6]
```diff
@@ -327,6 +327,12 @@ let one_command stop (ic,oc,id as conn) (backend:Backend.backend) =
       backend # last_entries2 i oc >>= fun () ->
       Lwt.return false
     end
+  | LAST_ENTRIES3 ->
+     Sn.input_sn ic >>= fun i ->
+      Logger.debug_f_ "connection=%s LAST_ENTRIES2: i=%Li" id i >>= fun () ->
```
Comment (domsj) [inline src/client/client_protocol.ml]: `LAST_ENTRIES3`

File: src/client/remote_nodestream.ml [hunk 7]
```diff
@@ -101,17 +101,18 @@ class remote_nodestream ((ic,oc) as conn) =
...
-            tlog_coll # save_tlog_file name length ic >>= fun () ->
-            loop_parts ()
-          end
+           begin
+             Logger.debug_f_ "save_file" >>= fun () ->
+             Llio.input_string ic >>= fun name ->
+             Sn.input_sn ic       >>= fun start_i ->
+             Llio.input_int64 ic  >>= fun length ->
+             Logger.info_f_ "retrieving %s (%Li bytes)" name length >>= fun () ->
+             tlog_coll # save_tlog_file name length ic >>= fun () ->
```
Comment (domsj) [inline src/client/remote_nodestream.ml]: `start_i` should be used here iso `name`
(also that means it's no longer needed to send `name` over the protocol)

Comment (toolslive) [inline src/client/remote_nodestream.ml]: Yes, I have the start_i, and it should be used, but we also need the name, as it determines where the file will end up (tlx dir or tlog dir).

File: src/main/replay_main.ml (hunk 5 above)
Comment (toolslive) [inline src/main/replay_main.ml:45]: This replay doesn't use a config. It feels a bit heavy to demand a full arakoon config just for this.

File: src/tlog/tlc2.ml [hunk 8]
```diff
@@ -725,164 +472,159 @@ class tlc2
...
+      (* what with rotation (jump to new tlog), open streams, ...*)
+      self # _maybe_close_current_file ()
+      >>= fun () ->                                                               
+      let full_name = TlogMap.get_full_path tlog_map remote_name in
+      let extension = Tlog_map.extension_of remote_name in
+      let tmp = full_name ^ ".tmp" in
+      
+      Logger.info_f_
+        "save_tlog_file: first_i:%s remote_name:%s in %s"
+        (Sn.string_of first_i) remote_name tmp
```
Comment (domsj) [inline src/tlog/tlc2.ml]: Hmm, so the tempfile name is based on the remote_name, which is kinda confusing. I think it would be clearer to use the remote name only to get the extension (see if it's a tlog/tlf/tlx), and base the temp file name on new_outer (as is done for the final filename below).

**Comment by domsj** [comment]
There's still a travis test failure https://s3.amazonaws.com/archive.travis-ci.org/jobs/108264111/log.txt

```
Failure: universe:11:system:2:startup:2:ahead_master_loses_role

value for 'invalidkey' should not be in store
```

**Comment by domsj** [comment]
ok
